rust/datafusion/src/physical_plan/unicode_expressions.rs

```diff
@@ -0,0 +1,537 @@
+                        .rev()
+                        .nth(n.abs() as usize - 1)
+                        .map_or("", |(i, _)| {
+                            &from_utf8(&string.as_bytes()[..i]).unwrap()
+                        }),
+                ),
+            },
+        })
+        .collect::<GenericStringArray<T>>();
+
+    Ok(Arc::new(result) as ArrayRef)
+}
+
+/// Extends the string to length length by prepending the characters fill (a space by default). If the string is already longer than length then it is truncated (on the right).
+/// lpad('hi', 5, 'xy') = 'xyxhi'
+pub fn lpad<T: StringOffsetSizeTrait>(args: &[ArrayRef]) -> Result<ArrayRef> {
+    match args.len() {
+        2 => {
+            let string_array = downcast_string_arg!(args[0], "string", T);
+            let length_array = downcast_arg!(args[1], "length", Int64Array);
+
+            let result = string_array
+                .iter()
+                .zip(length_array.iter())
+                .map(|(string, length)| match (string, length) {
+                    (None, _) => None,
+                    (_, None) => None,
+                    (Some(string), Some(length)) => {
+                        let length = length as usize;
+                        if length == 0 {
+                            Some("".to_string())
+                        } else {
+                            let graphemes = string.graphemes(true).collect::<Vec<&str>>();
+                            if length < graphemes.len() {
+                                Some(graphemes[..length].concat())
+                            } else {
+                                let mut s = string.to_string();
+                                s.insert_str(
+                                    0,
+                                    " ".repeat(length - graphemes.len()).as_str(),
+                                );
+                                Some(s)
+                            }
+                        }
+                    }
+                })
+                .collect::<GenericStringArray<T>>();
+
+            Ok(Arc::new(result) as ArrayRef)
+        }
+        3 => {
+            let string_array = downcast_string_arg!(args[0], "string", T);
+            let length_array = downcast_arg!(args[1], "length", Int64Array);
+            let fill_array = downcast_string_arg!(args[2], "fill", T);
+
+            let result = string_array
+                .iter()
+                .zip(length_array.iter())
+                .zip(fill_array.iter())
+                .map(|((string, length), fill)| match (string, length, fill) {
+                    (None, _, _) => None,
+                    (_, None, _) => None,
+                    (_, _, None) => None,
+                    (Some(string), Some(length), Some(fill)) => {
+                        let length = length as usize;
+
+                        if length == 0 {
+                            Some("".to_string())
```

Comment:
Is `String` necessary or could we use `&str` in those iterators? (to avoid an extra allocation per item)

Reply:
I'm hitting issues with the borrow checker so I think that the `String` allocation is required.